### PR TITLE
Fix networks tests

### LINKED ISSUE #161: Recover lost transformation data when fixing networks
The old results clean-up script removed the input components, so we need to use the old network to extract the phase of the calculations when working out which transformations worked.

### PR DESCRIPTION
**Pull request by jthorton**
## Summary of changes
Add tests for the fix networks script to cover most failure modes

## Licensing agreement
* [x] I agree to release these inputs under the combined MIT and CCBY SA 4.0 licenses of this repository

[templates]: https://github.com/OpenFreeEnergy/IndustryBenchmarks2024/tree/main/industry_benchmarks
[input_validation]: https://github.com/OpenFreeEnergy/IndustryBenchmarks2024/tree/main/industry_benchmarks/utils/input_validation.py

## TODO
- [x] Fix #161 
 - [x] Trigger the charge change warning while testing the fix network script 
~~- [ ] Fix the graphml annotations being lost during the round trip~~

### CODE REVIEW

File: industry_benchmarks/utils/fix_networks.py [hunk 1]
```diff
@@ -420,17 +444,17 @@ def get_fixed_alchemical_network(ducktape_network, alchemical_network):
     Create an alchemical networks with only the missing edges.
     """
     solv = openfe.SolventComponent()
-
+    cofactors = []
     for node in alchemical_network.nodes:
         prot_comps = [isinstance(comp, ProteinComponent) for comp in node.components.values()]
```
Comment (hannahbaumann) [inline industry_benchmarks/utils/fix_networks.py:449]: This is a bug (in the main script), it leads to not saving out the `ProteinComponent` and also not necessarily the cofactors. I added a suggestion in the main PR, could you maybe add a test for reading in one of the complex .json files that this script produces and check that it has `SmallMoleculeComponent`, `ProteinComponent`, solvent, and cofactor (if you use eg5)?

File: industry_benchmarks/utils/tests/test_fix_networks.py [hunk 2]
```diff
@@ -79,18 +124,125 @@ def test_parse_results(results, input_alchemical_network):
...
+        )
+        default_charge_settings = get_settings_charge_changes()
+        defaullt_settings = get_settings()
+        # make sure that charge change edges have the correct settings
+        for edge in new_network.edges:
+            charge_diff = get_alchemical_charge_difference(edge.mapping)
+            if charge_diff != 0:
+                assert edge.protocol.settings == default_charge_settings
+            else:
+                assert edge.protocol.settings == defaullt_settings
```
Comment (hannahbaumann) [inline industry_benchmarks/utils/tests/test_fix_networks.py:248]: Maybe add a check for the presence of the cofactor here?

**Review by hannahbaumann — CHANGES_REQUESTED** [review]
Thanks @jthorton , this looks great! Just left the two comments about checking for the presence of the expected components, either in the output alchemical network or the transformations itself would be great!

**Comment by hannahbaumann** [comment]
I also just realized the two cmet tests are failing for me, I think the folder cmet_results folder may be missing an `__init__.py` file?

File: industry_benchmarks/utils/fix_networks.py [hunk 3]
```diff
@@ -59,22 +60,43 @@ def _get_check_results_json(filename: str):
...
     """
-    # The way this was being done in PR #144 doesn't work because
-    # it would lead to either a solvent or vacuum simulation, not a
-    # complex one!
-    errmsg = ("The transformation data is not available, "
-              "please contact the OpenFE team")
-    raise ValueError(errmsg)
+    # grab the gufe key of the chemical systems used in the inputs
+    unit_result = list(pur["unit_results"].values())[0]
+    state_a_key = unit_result["inputs"]["stateA"][":gufe-key:"]
```
Comment (IAlibay) [inline industry_benchmarks/utils/fix_networks.py:72]: I thought we had del-ed the entire dictionary entry for these? https://github.com/OpenFreeEnergy/IndustryBenchmarks2024/blob/e376e39445118058e1904f86703a70ac1d468822/industry_benchmarks/utils/results_cleanup.py#L385

Comment (jthorton) [reply industry_benchmarks/utils/fix_networks.py:72]: I think the script removes the components in the `protocol_result` but leaves the key of the system in the `unit_results` which are stored in a different place. I generated the `bace_results/cleaned_results/complex_spiro10_spiro6.json` using the script in that commit to check it works but might have missed something. Do you have a copy of a cleaned directory I can test this on?

Comment (IAlibay) [reply industry_benchmarks/utils/fix_networks.py:72]: Ah yes, you're right! please ignore me

File: industry_benchmarks/utils/tests/test_fix_networks.py (hunk 2 above)
Comment (jthorton) [reply industry_benchmarks/utils/tests/test_fix_networks.py:266]: Good idea check added in [d550db5](https://github.com/OpenFreeEnergy/IndustryBenchmarks2024/pull/158/commits/d550db5308ba7070a5d95eff93f6a7f3fc504971)

**Comment by jthorton** [comment]
Thanks @hannahbaumann and @IAlibay ready for another pass!

File: industry_benchmarks/utils/tests/test_fix_networks.py [hunk 4]
```diff
@@ -79,18 +130,143 @@ def test_parse_results(results, input_alchemical_network):
...
+        # make sure that charge change edges have the correct settings
+        total_cofactors = 0
+        for edge in new_network.edges:
+            charge_diff = get_alchemical_charge_difference(edge.mapping)
+            if charge_diff != 0:
+                assert edge.protocol.settings == default_charge_settings
+            else:
+                assert edge.protocol.settings == defaullt_settings
+            # make sure the cofactor was added to the edge
+            if "protein" in edge.stateA.components:
```
Comment (hannahbaumann) [inline industry_benchmarks/utils/tests/test_fix_networks.py:268]: Could you add an assertion for a `ProteinComponent` in the component list here to ensure we have a `ProteinComponent`? This test will currently fail, we'll have to address the bug in the main PR.

File: industry_benchmarks/utils/tests/test_fix_networks.py [hunk 5]
```diff
@@ -79,18 +130,143 @@ def test_parse_results(results, input_alchemical_network):
...
+        default_charge_settings = get_settings_charge_changes()
+        defaullt_settings = get_settings()
+        # make sure that charge change edges have the correct settings
+        total_cofactors = 0
+        for edge in new_network.edges:
+            charge_diff = get_alchemical_charge_difference(edge.mapping)
+            if charge_diff != 0:
+                assert edge.protocol.settings == default_charge_settings
+            else:
+                assert edge.protocol.settings == defaullt_settings
```
Comment (hannahbaumann) [inline industry_benchmarks/utils/tests/test_fix_networks.py:266]:
```suggestion
                assert edge.protocol.settings == default_settings
```

File: industry_benchmarks/utils/tests/test_fix_networks.py [hunk 6]
```diff
@@ -79,18 +130,143 @@ def test_parse_results(results, input_alchemical_network):
...
+        # load the networks and check the protocol settings for each leg
+        output_alchemical_network = temp_out_dir / 'alchemical_network.json'
+        new_network = AlchemicalNetwork.from_dict(
+            json.load(
+                open(output_alchemical_network),
+                cls=JSON_HANDLER.decoder
+            )
+        )
+        default_charge_settings = get_settings_charge_changes()
+        defaullt_settings = get_settings()
```
Comment (hannahbaumann) [inline industry_benchmarks/utils/tests/test_fix_networks.py:258]:
```suggestion
        default_settings = get_settings()
```

**Review by hannahbaumann — CHANGES_REQUESTED** [review]
Thank you! Could you add that one additional check?

File: industry_benchmarks/utils/tests/test_fix_networks.py (hunk 4 above)
Comment (jthorton) [reply industry_benchmarks/utils/tests/test_fix_networks.py:268]: Good spot I thought it was strange the protein was False! Done in [2a0287c](https://github.com/OpenFreeEnergy/IndustryBenchmarks2024/pull/158/commits/2a0287c7ba260dbb990afb4fe7397b92c39aab3a)

**Review by hannahbaumann — APPROVED** [review]
Thanks @jthorton , LGTM!

**Pull request merged**